### ISSUE TITLE
GitHub Actions: Remove merged PRs from the Project Board

### ISSUE THREAD

**Issue by SAUMILDHANKAR** [hackforla/website#2592]
### Overview
As a developer, we have to ensure that our kanban board is organized for all teams so that productivity is high. For this issue, we want the cards of the merged pull requests to automatically be removed from the Project Board.

### Action Items

- [x] Review the notes below to understand the GitHub actions architecture we are currently moving towards
- [x] This should be a new job under the ["PR Trigger" workflow](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/pull-request-trigger.yml)
    - [x] This job should trigger whenever a PR is merged
    - [x] This job should have an if statement that triggers when the event name is "pull_request" and event action is "merged" 
        - if: ${{ github.event_name == 'pull_request' && github.event.action == 'merged'}}
    - [x] This job should run in parallel with other actions that may exist in the trigger file
  
### Pseudo-code

- [x] When a pull request is merged
    - [x] Remove the pull request from the project board.

### Checks

- [x] Test in your local environment that it works
- [x] If a reviewer merges a pr and does not remove the pr from the Project Board, the GitHub action should automatically remove the pr from the Project Board
- [x] If a reviewer merges a pr and moves it to the Project Board, the GitHub action should move it to the correct column in the Project Board and then remove it
- [x] If the column the pr is going to has automation turned on, the GitHub action should still trigger and remove the issue without causing any weird issues / side effects

### Resources/Instructions
Never done GitHub actions? [Start here!](https://docs.github.com/en/actions)
[GitHub Complex Workflows doc](https://docs.github.com/en/actions/learn-github-actions/managing-complex-workflows)
[GitHub Actions Workflow Directory](https://github.com/hackforla/website/tree/gh-pages/.github/workflows)
[Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)
[Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions)
[actions/github-script](https://github.com/actions/github-script)
[GitHub RESTAPI](https://docs.github.com/en/rest)

#### Architecture Notes
Currently, we want our GitHub Actions to be structured like so based on this [proposal](https://docs.google.com/spreadsheets/d/12NcZQoyGYlHlMQtJE2IM8xLYpHN75agb/edit#gid=1231634015):

- Schedules (military time)
        - Schedule Friday 0700
        - Schedule Thursday 1100
        - Schedule Daily 1100
- Linters
        - Lint SCSS
- PR Trigger
        - Add Linked Issue Labels to Pull Request
        - Add Pull Request Instructions
- Issue Trigger
        - Add Missing Labels To Issues
- WR - PR Trigger
        - WR Add Linked Issue Labels to Pull Request
        - WR Add Pull Request Instructions
- WR - Issue Trigger

Actions with the same triggers (excluding linters, which will be their own category) will live in the same github action file. Scheduled actions will live in the same file if they trigger on the same schedule (i.e. all files that trigger everyday at 11am will live in one file, while files that trigger on Friday at 7am will be on a separate file).

That said, this structure is not set in stone. If any part of it feels strange, or you have questions, feel free to bring it up with the team so we can evolve this format!

**Comment by harshitasao** [comment]
ETA : one week
Availability : Flexible

**Comment by harshitasao** [comment]
Progress : Going through the linked docs to make myself familiar with Github Actions.
Blockers : None so far.
Availability : Flexible.
ETA : Till next Sunday.

**Comment by harshitasao** [comment]
Progress: 4 task completed, Some are left.
Blockers: Have some clarification question.
Availability: Sunday (11am-1pm) IST
ETA: Till monday

**Comment by harshitasao** [comment]
Progress: 4 task completed, Some are left.
Blockers: Have some clarification questions.
Availability: Sunday (11am-1pm) IST
ETA: Till Wednesday

**Comment by harshitasao** [comment]
Progress: Almost done.
Blockers: Have some doubts.
Availability: Monday (3pm-5pm) IST
ETA: As quickly as possible.

**Comment by harshitasao** [comment]
Progress: not much progress.
Blockers: None.
Availability: Wednesday, Thursday, Friday 2 hours.
ETA: ASAP.

**Comment by tamara-snyder** [comment]
Hi @harshitasao, I noticed you haven't left a progress update on your issue recently. Please comment an update if you're still interested in working on this issue.

**Comment by ExperimentsInHonesty** [comment]
@harshitasao I am removing you from this issue because we have not heard from you since May.  If you are still working on it, you can reassign and move back into the inprogress.

**Comment by attali-david** [comment]
i. Availability: 9/4-9/6
ii. ETA: 9/6 EOD

**Comment by attali-david** [comment]
Blocker: I've read all the documentation referenced in the instruction, but I don't understand the documentation for the context actions/github-script. As described [here](https://github.com/actions/toolkit/blob/main/packages/github/src/context.ts), I don't see how I can access the project card ID in order to delete it. 

What I've done so far:

1. I created a yml file under .github/workflows based on the HackForLa guidelines found [here](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions). 
2. I created a script under github-actions/trigger-pr based on the same guidelines. 
3. I looked at other scripts in github-actions directory.
4. I read the context documentation [here](https://docs.github.com/en/actions/learn-github-actions/contexts) 

If someone could direct me to the right page or maybe some examples, I'd greatly appreciate it.

**Comment by blulady** [comment]
> Blocker: I've read all the documentation referenced in the instruction, but I don't understand the documentation for the context actions/github-script. As described [here](https://github.com/actions/toolkit/blob/main/packages/github/src/context.ts), I don't see how I can access the project card ID in order to delete it.
> 
> What I've done so far:
> 
>     1. I created a yml file under .github/workflows based on the HackForLa guidelines found [here](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions).
> 
>     2. I created a script under github-actions/trigger-pr based on the same guidelines.
> 
>     3. I looked at other scripts in github-actions directory.
> 
>     4. I read the context documentation [here](https://docs.github.com/en/actions/learn-github-actions/contexts)
> 
> 
> If someone could direct me to the right page or maybe some examples, I'd greatly appreciate it.

Again, it's late for me but let me know if [this](https://github.com/marketplace/actions/create-or-update-project-card) is not helpful...

**Comment by kathrynsilvaconway** [comment]
@attali-david 
Hi, this issue has not been updated in over 2 weeks. If you are still interested in working with HfLA, please let us know by updating your ETA and Availability. Also, plese reach out on Slack if you are still having a blocker, we can help. If we don't hear back within 7 days, we will unassign you from this issue and put it back in the queue. Thanks!

**Comment by blulady** [comment]
@attali-david 
Hey David, are you still working on this issue? If you are not please let me know and you can unassign yourself and find a different issue. If I don't hear back from you in a week, I'll assume you are done.
Thanks!!

**Comment by blulady** [comment]
Sending this back to New Issue Approval to see if anything needs to happen before moving it back to the prioritized backlog.

**Comment by t-will-gillis** [comment]
i. Availability: Tu, W, Th 2-5 pm
ii. ETA estimated: F 5 pm

**Comment by jdingeman** [comment]
Dev has reached out regarding this issue and is actively working on it.

@t-will-gillis, when you get a chance, please provide a progress update on this issue to note down your blockers and availability.

**Comment by t-will-gillis** [comment]
Progress: Have familiarized with GHAs have run test functions and set up test PRs/ MergePRs in own repo 
Blockers: I am unclear/ would like to have clarification about what is currently happening (both automatically and manually) with MergedPRs and what we would like to have happen with MergedPRs. 
Availability: Tuesday meeting, rest of week 11/29-12/4
ETA: Once unblocked, will be able to move to the next step of implementing a GHA. 
Pictures: n/a

**Comment by blulady** [comment]
@hackforla/product-management-maintainers @hackforla/product-managment @ExperimentsInHonesty 
"If a reviewer merges a pr and moves it to the Project Board, the GitHub action should move it to the correct column in the Project Board and then remove it" Need clarification what column the merged pr needs to be moved on to in the project board. What is the correct column in the Project Board that it needs to be moved to after? And when you say removed from the project board, do you mean deleted?

**Comment by JessicaLucindaCheng** [comment]
> @hackforla/product-management-maintainers @hackforla/product-managment @ExperimentsInHonesty "If a reviewer merges a pr and moves it to the Project Board, the GitHub action should move it to the correct column in the Project Board and then remove it" Need clarification what column the merged pr needs to be moved on to in the project board. What is the correct column in the Project Board that it needs to be moved to after? And when you say removed from the project board, do you mean deleted?

**Comment by ExperimentsInHonesty** [comment]
Hi Sarah @blulady the team `@hackforla/product-management-maintainers`  does not go only to hackforla website pms, it goes to all pms across all of HfLA.  Please don't use that.

I don't believe there is a PM @ messaging team on the website only.  You can use the ready for milestone, and we will read the comments every couple of days.

**Comment by ExperimentsInHonesty** [comment]
## Documentation of the Current Automation

## Columns

### New Issue Approval
- Automation Name:  To Do
  - Move issues here when…
     - Newly added: Issues will automatically move here when added to this project.
     - Reopened: If a closed issue in this project reopens, it will automatically move here.
### QA
- Automation Name: Done
  - Move issues here when…
    - Closed: If an open issue in this project is closed, it will automatically move here.
### PR Needs review (Automated Column, do not place items here manually)
- Automation Name: In Progess
  - Move pull requests here when…
    - Newly added:  Pull requests will automatically move here when added to this project.
    - Reopened: If a closed pull request in this project reopens, it will automatically move here.
### test-pending-approval (Automated Column, do not place items here manually)
- Automation Name: In Progress
  - Move pull requests here when…
    - Pending approval by reviewer: Pull requests in this project will automatically move here when a reviewer requests changes, or it no longer meets the minimum number of required approving reviews. Recommended when another column has the Approved by reviewer automation enabled.
### test-approved-by-reviewer (Automated Column, do not place items here manually)
- Automation Name: In Progress
  - Move pull requests here when…
     - Approved by reviewer: Pull requests in this project will automatically move here when they meet the minimum number of required approving reviews. Recommended when another column has the Pending approval by reviewer automation enabled.

**Comment by ExperimentsInHonesty** [comment]
@blulady   I didn't write this issue, but I think the idea was to reduce the number of items on the board.  However, we want to keep the PRs for the feature branch that won't be merged 

So I just went through the [test-approved-by-reviewer (Automated Column, do not place items here manually) - with is:closed](https://github.com/hackforla/website/projects/7?card_filter_query=is%3Aclosed#column-14938060) column with filter on, and moved the 12 issues that should have been in the `Feature Branch - don't merge into gh-pages branch yet`

That leaves us with 148 issues that we could remove from the board since they are closed.  Which would allow us to see the 1 that the merge team actually needs to merge [test-approved-by-reviewer (Automated Column, do not place items here manually) - with is:open](https://github.com/hackforla/website/projects/7?card_filter_query=is%3Aopen#column-14938060)

I also manually moved a bunch of issues #3667 and #3536, etc from the `PR Needs review (Automated Column, do not place items here manually)` column where they should not have been because they are already closed without merging and they go into the ignore column.

That still leaves us with 5 of the 6 issues in the  `PR Needs review (Automated Column, do not place items here manually)` column that are merged already and should not be in the column.  Theoretically, they should be showing up in the `test-approved-by-reviewer (Automated Column, do not place items here manually)` with the other merged changes. So that people looking for PRs to review could look and see the 1 issue that does need a review.

Let me know if you need more clarification

**Comment by t-will-gillis** [comment]
Edited:  see update below

~~I have been reviewing some of the other issues on the Project Board related to GitHub Actions, and I would like to ask for some additional clarification on this issue in light of the others. Regarding *only* ‘merged’ PRs, what exactly do we want to have happen with them? Do we want them all to be removed completely, or all of them to be moved to a specific column (‘Done’, ‘QA’, something else?), or is there a set of rules to be used for which issues to move to which columns and for which reasons?~~
~~First, my key takeaways from this issue are:~~
1.	~~This issue says that ‘merged’ PRs are to be removed from the board- and not moved to any other column (except possibly one). I make a long argument below about this, but it comes down to: there are no appropriate columns to move a ‘merged’ PR to (except possibly one).~~ 
2.	~~I am belaboring the point, but this issue is specifically for what to do with ‘merged’ PRs. There are at least two other open issues discussing ‘open’/’opened’ PRs- the movements of ‘open’ PRs through the Project Board are not applicable to this issue. (see below)~~  
1.	~~Regarding the removal of ‘merged’ PRs:
•	There is an old issue , itself ironically closed but not addressed, that was discussing which automations to run to do which actions with closed issues.
•	The comments explain why this issue was initially closed without action, and why this issue is moving to ‘New Issue Approval’- It sounds like the question of what to do with closed issues is not resolved yet 
•	Since ‘merged’ == ‘closed’ in GitHub, what exactly do we want to do with ‘merged’/’closed’ PRs, and what bearing does this have on the unresolved-soon-to-be-reopened issue ?~~
2.	~~Regarding ‘open’/’opened’ PRs:
•	The action that Justin is working on is for which actions to take on open PRs, references moving them to the ‘correct’ columns for various scenarios when the PR is first ‘opened’.
•	Open issue  is also for ‘opened’ PRs, mentions creating new columns, mentions moving the open PRs to ‘correct’ columns, and mentions the merge team moving ‘merged PRs’ out of columns into other columns (but not where or why), and then deleting them.
•	The current issue is talking only about ‘merged PRs’, and moving those ‘merged PRs’ to and from ‘correct’ columns before deleting the issues. But in issue  the ‘correct’ columns are all for 'opened' PRs. The columns listed are definitely not for 'merged' PRs:
	PRs that need a reviewer assigned 
	PRs that need actions from authors 
	Ready for re-review 
	Reviewer has approved, waiting for merge team to merge 
•	From Bonnie’s response above, this leaves the ‘QA’ column as a possibly appropriate place for (some/all?) ‘merged’ PRs to move to. If all ‘merged’ PRs move here, then we won’t be removing them from the board. If some ‘merged’ PRs move here, we will need rules for which ones move here and which ones are removed completely. Maybe none move here?~~

**Comment by t-will-gillis** [comment]
@blulady @jdingeman @arpitapandya 
1. Progress: 

a.	I was able to create a model the HfLA ‘Project Board’ incl. matching columns in my repo. The goal was to see if HfLA’s GHAs/ *.yml files would be triggered by appropriate actions on the model using a personal access token where needed. This experiment was partially successful, in that the actions/ *.yml files attempted to run when issues and PRs were created, moved, closed, and merged.  

b.	I researched how to delete an issue when a PR is merged. In HfLA’s existing `move-closed-issues.yaml` the marketplace automation `alex-page/gihub-project-automation-plus@v0.8.1` *should* easily allow this therefore I added the required syntax to the `pull-request-trigger.yml`. 

c.	Unfortunately, when the automation attempts to run, neither the project (named here ‘Project Board’) nor the column can be found in my repository. This is a known issue that other users have reported. However, this issue has not been occurring in HfLA’s workflows for `move-closed-issues.yaml`. 

d.	My working theory is this: 
- HfLA’s uses the ‘Projects (classic)’ for its ‘Project Board’. Since Github discontinued ‘Projects (classic)’, I am using the new Projects and I believe this is leading to some incompatibilities with the `alex-page` automations.
-  I have looked at the ‘GIT API’ for how to automatically delete an issue. I am not certain what is possible here yet- but if the `alex-page` automation works with HfLA’s ‘Project Board’ this seems much more straightforward. 
- I have explored the Community comments at Github and it seems that many people are noticing a loss of GHA features changing from the ‘Projects (classic)’ to the new ‘Projects.’
- Regardless, I feel that it will be difficult to compare whether feature edits that work on the new ‘Projects’ will or won’t work on the ‘Projects (classic)’, and vice versa.

e.	My proposed solution: 
- I would like to test edits to the `pull-request-trigger.yml` file on the actual HfLA ‘Project Board’. I propose to add a new ‘Dummy’ column to the board, place one merged PR in this column, and run the Pull Request Trigger to see whether or not the card is deleted. 
- If this works, then the Pull Request Trigger could be edited to include the same of the three PR columns [ ‘PR Needs review (Automated Column…)’, ‘test-pending-approval (Automated Column…)’, and ‘test-approved-by-reviewer (Automated Column…)’ ]
-  I am very open to other suggestions as well. 

2.	Blockers: As I said, I am open to ideas but at this moment I believe I am at an impasse, and that I need access to either the actual ‘Project Board’ or a model ‘Project Board’ created with Github’s ‘Projects (classic)’. See above for my proposed solution for becoming unblocked. 
3.     Availability: Weekend of 6-8 January 23.
4.	ETA: tbd
5.	Pictures:

**Comment by t-will-gillis** [comment]
1. Progress: No direct progress from last update: I still plan to explore using the `alex-page/gihub-project-automation-plus@v0.8.1` automation for deleting `closed` + `merged` issues- the documentation indicates this should be possible. The HfLA GHA `move-closed-issue` also uses this automation. The logs show that up until a month ago, `move-closed-issue` was working. Currently the action fails with an error message 'bad credentials'. I am assuming that the HfLA token associated with the automation needs to be updated/reset- A member of the merge team will be updating this token in the last couple of days.
>>Once the token is updated/reset, and (hopefully) `move-closed-issue` is running again, THEN this will be a green light to continue exploring whether the `alex-page` automation can be used to delete `closed` + `merged` issues. 
2. Should be able to continue if the existing action begins working again with the revised token.
3. 5-6 hours available over the next week.
4. tbd
5. n/a

**Comment by t-will-gillis** [comment]
1. Progress: As of an hour ago, the HfLA GHA `Move Closed Issues` is working again. (yay) Thank you to whomever reset/updated the token/secrets. With the action running successfully this is confirmation that the `alex-page/gihub-project-automation-plus@v0.8.1` automation is also working. 

Please see the above notes for what I have done so far. What I would like to try now is to add a column "TEST-2592" to the project board, and a test file "TEST Pull Request Trigger 2592", then test on the HfLA website.
